### PR TITLE
Add AWS S3 Image Hosting, Litmus Tests and Build Zipping

### PR DESCRIPTION
**Pull request by iDVB**
- Replace `$.injectString` with `$.replace`
- Refactor `/dist/assets/img` to `/dist/images` to match typical email vendor requirements for zips
- Add `example.config.json` to illustrate usage
- Add task `npm run litmus`  to create litmus tests (target devices/clients) on account (config.json)
    - Add function to push all images to S3 account (config.json)
- Add task `npm run zip` to create individual zips (images + html) for each email

**Suggested Next Steps:**
- Update documentation
- Implement helpers to automated deploy to email vendors such as Silverpop
- Clean up [how account creds get stored](http://docs.aws.amazon.com/AWSJavaScriptSDK/guide/node-configuring.html#Credentials_from_Environment_Variables).

### CODE REVIEW

File: gulpfile.babel.js [hunk 1]
```diff
@@ -61,9 +74,9 @@ function sass() {
 
 // Copy and compress images
 function images() {
-  return gulp.src('src/assets/img/*')
+  return gulp.src('src/assets/images/*')
```
Comment (kball) [inline gulpfile.babel.js]: What's the driver behind changing directory names?  I'd rather maintain backwards consistency wherever possible

Comment (iDVB) [inline gulpfile.babel.js]: Good call. Originally it was to align to what I thought was needed in the zip task (vendor requirement) however, I now know the images don't even need to be in a folder and as such, if they are the only requirement is that the html refs relatively. 

So I'll revert the dir name change back.  Anything else? Should I be PRing the master branch? Are we at the point were I should include read me updates?

Comment (kball) [inline gulpfile.babel.js]: Yeah I checked with @rafibomb and we're not really versioning the template at this time, so PR against master is fine.  I'm going to try to carve out some time later today to actually test drive this, but on a readthrough it looked pretty good, so yeah lets go ahead and add readme updates as well.

Comment (iDVB) [inline gulpfile.babel.js]: Will do. Reverting the img directory changes and then updating the readme.

**Comment by iDVB** [comment]
Last two commits should have corrected the directories and updated the README.md

**Comment by kball** [comment]
Awesome.  @tdhartwick can you take this for a spin?

**Comment by tdhartwick** [comment]
@iDVB This is so awesome!! This is going to make a huge impact on people's workflow! Seriously, the Foundation Team is super pumped over here.

A few things we're noticing before we merge this in:

- [ ] **Merge-stream in package.json**. Looks like this package wasn't included into the package.json as a dependency.
- [ ] **Checklist Subject Lines**. I'm not quite following lines 2-4 for `config.json`. It doesn't seem to be applying a subject line to our checklists. If you have multiple emails you're working on, Litmus keeps replacing the checklist until it's reached the last email. Here's how we've approached it:
    - Added the front matter `--- subject: My Email Subject ---` (with the different content for each title) to each page.
    - Updated `<title>{{subject}}</title>` to the default layout: `default.html`
We've found that this creates separate checklists with the corresponding title tag as the checklist name in Litmus, instead of the default date.
- [ ] **Versions in Checklists?** This is a nice-to-have, but with your experience with the Litmus API is there a way to create new versions of a checklist if the subject line is the same? The current interaction is that the checklist is replaced with an updated checklist.
- [ ] **Silently failing if no AWS credentials are loaded**. We removed the attributes of the `aws` object in the `config.json`. We still had images in our src email. It failed and stopped executing the command. Is it possible to provide a warning in the terminal and continue with the litmus gulp task?

Again, thanks! This is huge.

**Comment by iDVB** [comment]
I haven't forgot about this. I have most of the mentioned fixes in. 
I just need to tie off the AWS+Litmus credential checks and I'll be able to commit again.
I'll get something in before the weekend is out, likely sooner.

**Comment by tdhartwick** [comment]
Awesome! Thanks @iDVB!

**Comment by iDVB** [comment]
@tdhartwick 
OK, So I added almost everything you mentioned. 

- Added Merge-stream to package.json
- Added Checklist Subject Lines implemented using front matter
- Refactored AWS creds object to match awspublish
- Fix, No error if AWS credentials are not given in the JSON. You can now optionally supply them any number of ways by following the [aws-sdk suggestions](http://docs.aws.amazon.com/AWSJavaScriptSDK/guide/node-configuring.html)
- Updated README.md 

**Not added:**
- Versioned Checklist - Not entirely convinced this wouldn't make things more confusing if you we're running a number of tests and just wanted to see the most recent one.  However, maybe this could be done by injecting the [git hash](https://www.npmjs.com/package/git-rev) or a random hash just after the `{{subject}}`? Personally, I like the idea of the git hash more. This way... you can archive the tests and roll back to the commit that produced them.

```html
<title>{{subject}} {{rev}}</title>
```

Let me know when you go ahead and merge the PR. Next step for me is to go over to [rwdemail](https://github.com/iDVB/rwdemail) and deprecate it and point to this. :P

**Comment by rafibomb** [comment]
@iDVB This is really awesome!

We are planning v2.1 to be released on Friday. Would love to get this in and add your contribution to the blog post 💪

@tdhartwick Are we good to go on this?

**Comment by kball** [comment]
Tried this out, it looks amazing.  Merged.  Thank you for the great work @iDVB

**Comment by iDVB** [comment]
Great! Thanks everyone.
It was good fun working on it. I'm a huge supporter of automation workflows/tools. Especially around areas of significant challenge like email. There are some other features that might be really handy.

**Some other ideas:**
Automated uploading to Email Marketing Vendors (Silverpop, Campaign Monitor, MailChimp etc)

I have seen this done even where there is no API, by essentially using a [headless browser](http://phantomjs.org/) and script to click through the UI. Considering this step is usually done a few times and about a 5-10min process, automating it across all emails could really speed things up.

Anyway, keep me in mind if I can be of any more help. I think the work you guys are doing with Foundation is great.

**Comment by rafibomb** [comment]
Hi @iDVB 
Email me - rafi at zurb .com for a Yeti shirt! Also, want to add you into our blog post on this fro tomorrow!

**Comment by rafibomb** [comment]
Those other features are great too! There is so much that can be done to make the emails workflow faster. 

I added these things to the Roadmap and we'd love this to be a reality as well.

**Comment by iDVB** [comment]
Hey @rafibomb email is already out. Maybe check junk folder? It would be coming from dan at danvanbrunt .com  Ping me again if you don't get it.

Thanks again!

File: gulpfile.babel.js [hunk 2]
```diff
@@ -154,10 +160,52 @@ function litmus() {
...
+    var fileName = path.basename(sourcePath, ext);
+
+    var moveHTML = gulp.src(sourcePath)
+      .pipe($.rename(function (path) {
+        path.dirname = fileName;
+        return path;
+      }));
+
+    var moveImages = gulp.src(sourcePath)
+      .pipe($.htmlSrc({ selector: 'img'}))
```
Comment (joefraley) [inline gulpfile.babel.js]: I'd love this to work. In my workflow I'm seeing gulp crash at this point. 

npm run zip

Two console logs here to see where the crash occurs. See the first one succeed in the output above.

gulpfile

Never reach the second log, so the crash occurs in the `gulp.src` or `$.htmlSrc` functions?

error output

Not sure where to file an issue for this. This commit seemed like the most logical place.

Comment (iDVB) [inline gulpfile.babel.js]: Can you please move this to an actual "Issue" and feel free to reference this commit if you feel there is a correlation.

**Pull request merged**